README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

Maze builder using a Dart implementation of Eller's Algorithm, as described here: http://www.neocomputer.org/projects/eller.html

This algorithm creates 'perfect' mazes, which guarantees a single path between any two cells, such as:
```
+---+---+---+---+---+---+---+
|           |           |   |
+---+   +---+   +   +   +   +
|   |   |       |   |       |
+   +   +   +   +   +   +   +
|       |   |   |   |   |   |
+   +---+   +   +---+---+   +
|   |   |   |   |   |   |   |
+   +   +   +   +   +   +   +
|   |       |   |   |       |
+   +---+   +---+   +---+---+
|   |   |   |       |       |
+   +   +   +   +---+   +   +
|                       |   |
+---+---+---+---+---+---+---+
```

This library generates a two-dimensional array of maze cells, each with the following properties:

    {
      x: 4,          // Horizontal position, integer
      y: 7,          // Vertical position, integer
      top: false,    // Top/Up has a wall/blocked if true, boolean 
      left: false,   // Left has a wall/blocked if true, boolean
      bottom: true,  // Bottom/Down has a wall/blocked if true, boolean
      right: true,   // Right has a wall/blocked if true, boolean
      set: 5         // Set # used to generate maze, can be ignored
    }

## Features

The example of this package also draws the lines using Flutter CustomPainter.

## Getting started

add the library `import 'package:maze_builder/maze_builder.dart';`

## Usage


```dart
Random rand = Random();
List<List<Cell>> localMaze = generate(width: 200, height: 200, closed: true, seed: rand.nextInt(100000));
```

Receive a `Cell` array of all the points of the map

Then draw them

```dart
makeBlocks() {
    int mazeLength = maze.length;
    for (var i = 0; i < mazeLength; i++) {
      for (var j = 0; j < maze[i].length; j++) {
        getBlockLines(maze[i][j], i);
      }
    }
  }

  getBlockLines(Cell cell, int index) {
    
    int size = 24;
    if (cell.top) {
      double x1 = cell.x * size;
      double y1 = cell.y * size;
      double x2 = cell.x * size + size;
      double y2 = cell.y * size;
      drawLine(x1, y1, x2, y2, _paint);
    }
    if (cell.bottom) {
      double x1 = cell.x * size;
      double y1 = cell.y * size + size;
      double x2 = cell.x * size + size;
      double y2 = cell.y * size + size;

      drawLine(x1, y1, x2, y2, _paint);
    }
    if (cell.left) {
      double x1 = cell.x * size;
      double y1 = cell.y * size;
      double x2 = cell.x * size;
      double y2 = cell.y * size + size;

      drawLine(x1, y1, x2, y2, _paint);
    }
    if (cell.right) {
      double x1 = cell.x * size + size;
      double y1 = cell.y * size;
      double x2 = cell.x * size + size;
      double y2 = cell.y * size + size;

      drawLine(x1, y1, x2, y2, _paint);
    }
  }

    // DRAW LINE ///////////////////////////////
  void drawLine(double x, double y, double targetX, double targetY, Paint paint) {

    canvas!.drawLine(
        Offset(x, y),
        Offset(targetX, targetY),
        paint,
    );

   }

  ```


## Additional information

This library is a direct Dart port of https://github.com/bestguy/generate-maze


maze_builder/CHANGELOG.md
## 1.0.1

* Added correct formatting and Documentation comments
## 1.0.0

* Initial release, added example and ReadMe


maze_builder/test/maze_builder_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:maze_builder/maze_builder.dart';

void main() {
  test('adds one to input values', () {
    final maze = generate();
    expect(maze.length, maze.isNotEmpty);
  });
}


maze_builder/lib/maze_builder.dart
class Cell {
  double x;
  double y;
  bool top;
  bool left;
  bool right;
  bool bottom;
  double? set;

  /// The cell model class
  Cell({
    required this.x,
    required this.y,
    required this.top,
    required this.left,
    required this.bottom,
    required this.right,
    this.set,
  });

  getPropertyByKey(String key) {
    switch (key) {
      case "set":
        {
          return set;
        }

      case "x":
        {
          return x;
        }

      case "y":
        {
          return y;
        }

      case "top":
        {
          return top;
        }
      case "bottom":
        {
          return bottom;
        }
      case "left":
        {
          return left;
        }
      case "right":
        {
          return right;
        }
    }
  }
}

/// [internal function] make a list based on a predicate
List<dynamic> compact(List<dynamic> array) {
  return array.where((u) => u != 0.0).toList();
}

/// [internal function] difference between two lists
List<dynamic> difference(List<dynamic> c, List<dynamic> d) {
  return [c, d].reduce((a, b) => c.toSet().difference(b.toSet()).toList());
}

List<dynamic> initial(List<dynamic> array) {
  return array.sublist(0, array.length - 1);
}

/// { [key]: T[] }
Map<int, List<dynamic>> oldgroupBy(List<Cell> list, String key) {
  List<dynamic> keys = list.map((item) => item.getPropertyByKey(key)).toList();
  var uniqKeys = uniq(keys).toList();
  Map<int, List<dynamic>> _dict = {};
  uniqKeys.asMap().forEach((index, element) {
    _dict[element.toInt()] = [];
  });

  // uniqKeys.reduce((prev, next) => {
  //       {
  //         ...prev,
  //         [next]: []
  //       }
  //     });
  Map<int, List<dynamic>> dict = _dict;

  list.forEach((item) => dict[item.getPropertyByKey(key)]?.add(item));
  //print(dict);
  return dict;
}

/// [internal function] group iterables by predicate
Map<T, List<S>> groupBy<S, T>(Iterable<S> values, T Function(S) key) {
  var map = <T, List<S>>{};
  for (var element in values) {
    (map[key(element)] ??= []).add(element);
  }
  return map;
}

/// [internal function] get the last element of a list
last(List<dynamic> array) {
  return array[array.length - 1];
}

/// [internal function] make a list based on predicate
Iterable range(double n, {int end = 0}) {
  return end != 0
      ? List.from(List.generate(end - n.round(), (int index) => index,
              growable: true))
          .map((k) => k + n)
      : List.from(
          List.generate(n.round(), (int index) => index, growable: true));
}

/// [internal function] return a duplicate free list
List<dynamic> uniq(List<dynamic> array) {
  return [...new Set.from(array)];
}

/// [internal function] return a list with random elements
List<dynamic> sampleSize(List<dynamic>? array, double? n, Function random) {
  n = n ?? 1;
  double length = (array == null) ? 0.0 : array.length.toDouble();

  /// !length note
  if (length == 0 || n < 1) {
    return [];
  }
  n = n > length ? length : n;
  int index = -1;
  double lastIndex = length - 1;
  List result = [...?array];
  while (++index < n) {
    //var _random = random();
    var rand = index + (random() * (lastIndex - index + 1)).floor() as int;
    Cell value = result[rand];
    result[rand] = result[index];
    result[index] = value;
  }
  return result.sublist(0, n.round());
}

/// [internal function] randomization engine
Function mulberry32(int seed) {
  return () {
    seed += 0x6D2B79F5;
    int t = seed;
    t = imul2(t ^ t >> 15, t | 1);
    t ^= t + imul2(t ^ t >> 7, t | 61);
    double result = ((t ^ t >> 14) >> 0) / 4294967296;

    return result;
  };
}

/// [internal function] imul2 implementation in Dart
int imul2(int a, int b) {
  int aHi = (a >> 16) & 0xffff;
  int aLo = a & 0xffff;
  int bHi = (b >> 16) & 0xffff;
  int bLo = b & 0xffff;
  // the shift by 0 fixes the sign on the high part
  // the final |0 converts the unsigned value into a signed value
  return ((aLo * bLo) + (((aHi * bLo + aLo * bHi) << 16) >> 0)).toSigned(32);
}

/// [internal function] merge a list with another
List<Cell> mergeSetWith(List<Cell> row, double oldSet, double newSet) {
  row.forEach((box) {
    if (box.set == oldSet) box.set = newSet;
  });

  return row;
}

/// [internal function] add missing set to list
List<Cell> populateMissingSets(List<Cell> row, Function random) {
  var _map = row.map((row) => row.set).toList();
  List<dynamic> _uniq = uniq(_map);
  List<dynamic> setsInUse = compact(_uniq);
  List<dynamic> allSets = range(1, end: row.length + 1).toList();
  List<dynamic> diff = difference(allSets, setsInUse);
  List<double> availableSets = diff.cast<double>();
  availableSets.sort((a, b) => (0.5 - random()).sign.toInt());
  // print("$availableSets, $setsInUse, $allSets");
  row.where((box) => box.set == 0).toList().asMap().forEach((i, box) {
    box.set = availableSets[i];
  });

  return row;
}

/// [internal function] merge randomized sets into a new list
List<Cell> mergeRandomSetsIn(List<Cell> row, Function random,
    {probability = 0.5}) {
  // Randomly merge some disjoint sets

  var allBoxesButLast = initial(row);
  allBoxesButLast.asMap().forEach((x, current) {
    var next = row[x + 1];
    var differentSets = current.set != next.set;
    var rand = random();
    var shouldMerge = rand <= probability;

    if (current.x == 0 && current.y == 0) {
      //print(">>>> $rand  $differentSets $shouldMerge ${current.set}, ${next.set}");
    }
    if (differentSets && shouldMerge) {
      row = mergeSetWith(row, next.set!, current.set);
      current.right = false;
      row[x + 1].left = false;
    }
  });

  allBoxesButLast.add(row[row.length - 1]);
  return allBoxesButLast as List<Cell>;
}

/// [internal function] randomly add an exit to each set
addSetExits(List<Cell> row, List<Cell> nextRow, Function random) {
  // Randomly add bottom exit for each set
  List<dynamic> setsInRow = [];

  groupBy(row, (item) => (item).set!.round()).forEach((key, value) {
    setsInRow.add(value);
  });

  setsInRow.forEach((set) {
    List<dynamic> exits =
        sampleSize(set, (random() * set.length).ceil().toDouble(), random);
    exits.forEach((exit) {
      //if (exit) {
      Cell below = nextRow[exit.x.round()];
      exit.bottom = false;
      below.top = false;
      below.set = exit.set;
      //}
    });
  });

  return setsInRow;
}

/// Generate a maze (List of List of Cell objects)
List<List<Cell>> generate(
    {int width = 8, int height = 0, closed = true, seed = 1}) {
  height = width;

  Function random = mulberry32(seed);
  List<List<Cell>> maze = [];
  var r = range(width.toDouble());

  // Populate maze with empty cells:
  for (var y = 0; y < height; y += 1) {
    var row = r.map((x) {
      return Cell(
        x: x.toDouble(),
        y: y.toDouble(),
        top: closed || y > 0,
        left: closed || x > 0,
        bottom: closed || y < (height - 1),
        right: closed || x < (width - 1),
        set: 0,
      );
    }).toList();
    maze.add(row);
  }

  // All rows except last:
  initial(maze).asMap().forEach((y, row) {
    row = populateMissingSets(row, random);
    row = mergeRandomSetsIn(row, random);
    addSetExits(row, maze[y + 1], random);
  });

  var lastRow = last(maze).toList();
  lastRow = populateMissingSets(lastRow, random);
  lastRow = mergeRandomSetsIn(lastRow, random, probability: 1);

  return maze;
}


maze_builder/example/lib/maze_drawer.dart
import 'dart:core';
import 'dart:math';

import 'package:flutter/rendering.dart';
import 'package:flutter/material.dart';
import 'package:maze_builder/maze_builder.dart';
import 'dart:ui' as ui;

class MazeDrawer {
  Color color = const Color.fromARGB(255, 254, 238, 13);
  Canvas? canvas;
  Paint _paint = Paint();
  double radius = 10.0;
  int fps = 24;
  int printTime = DateTime.now().millisecondsSinceEpoch;
  final _random = Random();
  int blockSize = 8;
  List<List<Cell>> maze = [];
  ui.BlendMode? blendMode = ui.BlendMode.src;
  bool isMazeDrawn = false;
  final Rect _bounds = const Rect.fromLTWH(0, 0, 0, 0);

  bool maxRightReached = false;

  /// Constructor
  MazeDrawer({
    required this.maze,

    /// <-- The delay until the animation starts
    required this.blockSize,
    solution,

    /// <-- The particles blend mode (default: BlendMode.src)
    this.blendMode,
  }) {
    /// default painter

    _paint = Paint()
      ..strokeCap = StrokeCap.round
      ..isAntiAlias = true
      ..color = color
      ..strokeWidth = 2.0
      ..strokeJoin = StrokeJoin.round
      ..style = PaintingStyle.fill;
  }

  void update(Canvas canvas) {
    this.canvas = canvas;
    makeBlocks();
  }

  makeBlocks() {
    int mazeLength = maze.length;
    for (var i = 0; i < mazeLength; i++) {
      for (var j = 0; j < maze[i].length; j++) {
        getBlockLines(maze[i][j], i);
      }
    }

    isMazeDrawn = true;
  }

  getBlockLines(Cell cell, int index) {
    //var lines = [];
    //print(cell);
    int size = blockSize;
    if (cell.top) {
      double x1 = cell.x * size;
      double y1 = cell.y * size;
      double x2 = cell.x * size + size;
      double y2 = cell.y * size;
      drawLine(x1, y1, x2, y2, _paint);
    }
    if (cell.bottom) {
      double x1 = cell.x * size;
      double y1 = cell.y * size + size;
      double x2 = cell.x * size + size;
      double y2 = cell.y * size + size;

      drawLine(x1, y1, x2, y2, _paint);
    }
    if (cell.left) {
      double x1 = cell.x * size;
      double y1 = cell.y * size;
      double x2 = cell.x * size;
      double y2 = cell.y * size + size;

      drawLine(x1, y1, x2, y2, _paint);
    }
    if (cell.right) {
      double x1 = cell.x * size + size;
      double y1 = cell.y * size;
      double x2 = cell.x * size + size;
      double y2 = cell.y * size + size;

      drawLine(x1, y1, x2, y2, _paint);
    }
  }

  void drawCircle(double x, double y, Paint paint) {
    updateCanvas(0, 0, () {
      canvas!.drawCircle(Offset(x, y), radius, paint);
    });
  }

  // DRAW LINE ///////////////////////////////
  void drawLine(double x, double y, double targetX, double targetY, Paint paint) {
    //Rect bounds = _camera!.getCameraBounds();

    updateCanvas(_bounds.left * -1, _bounds.top, () {
      //print("$x, $y, $signX, $signY");
      canvas!.drawLine(
        Offset(x, y),
        Offset(targetX, targetY),
        paint,
      );
    }, translate: true);
  }

  Rect rect() => Rect.fromCircle(center: Offset.zero, radius: radius);

  double doubleInRange(double start, double end) {
    if (start == end) {
      return start;
    } else {
      return _random.nextDouble() * (end - start) + start;
    }
  }

  double randomDelay({double min = 0.005, double max = 0.05}) {
    if (min == max) {
      return min;
    } else {
      return doubleInRange(min, max);
    }
  }

  void delayedPrint(String str) {
    if (DateTime.now().millisecondsSinceEpoch - printTime > 10) {
      printTime = DateTime.now().millisecondsSinceEpoch;
      //print(str);
    }
  }

  void updateCanvas(double? x, double? y, VoidCallback callback, {bool translate = false}) {
    double localX = x ?? 0;
    double localY = y ?? 0;
    canvas!.save();
    if (translate) {
      canvas!.translate(localX, localY);
    }
    //canvas!.translate(0, 0);
    callback();
    canvas!.restore();
  }
}


maze_builder/example/lib/maze_driver.dart
import 'dart:core';
import 'package:example/maze_drawer.dart';
import 'package:flutter/material.dart';
import 'package:maze_builder/maze_builder.dart';
import 'dart:ui' as ui;
import 'package:vector_math/vector_math.dart' as vectorMath;

class MazeDriverCanvas extends CustomPainter {
  List<Map<String, dynamic>> points = [];

  Color color = Colors.black;
  var index = 0;
  var offset = 0;
  AnimationController? controller;
  Canvas? canvas;
  int delay = 500;
  int currentTime = 0;
  int fps = 24;
  int printTime = DateTime.now().millisecondsSinceEpoch;
  int timeDecay = 0;
  double rate = 1.0;
  int timeToLive = 24;
  double width = 100;
  double height = 100;
  int curveIndex = 0;
  double computedAngle = 0.0;
  List<List<vectorMath.Vector2>> curve = [];
  Function? update;
  Paint _paint = Paint();
  int blockSize = 8;
  BoxConstraints sceneSize = const BoxConstraints(
      minWidth: 800, maxWidth: 1600, minHeight: 450, maxHeight: 900);
  ui.BlendMode? blendMode = ui.BlendMode.src;

  bool maxRightReached = false;
  List<List<Cell>> maze = [];
  MazeDrawer? mazeMap;
  //

  /// Constructor
  MazeDriverCanvas({
    /// <-- The animation controller
    required this.controller,

    /// <-- The delay until the animation starts
    required this.width,
    required this.height,
    required this.blockSize,
    required this.maze,

    /// <-- The particles blend mode (default: BlendMode.src)
    this.blendMode,
  }) : super(repaint: controller) {
    /// the delay in ms based on desired fps
    timeDecay = (1 / fps * 1000).round();
    _paint = Paint()
      ..strokeCap = StrokeCap.round
      ..isAntiAlias = true
      ..color = Colors.yellow
      ..strokeWidth = 2.0
      ..strokeJoin = StrokeJoin.round
      ..style = PaintingStyle.stroke;

    /// make maze
    mazeMap = MazeDrawer(maze: maze, blockSize: blockSize);
  }

  @override
  void paint(Canvas canvas, Size size) {
    this.canvas = canvas;

    paintImage(canvas, size);

    mazeMap!.update(canvas);
  }

  void paintImage(Canvas canvas, Size size) async {
    draw(canvas, size);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return false;
  }

  void draw(Canvas canvas, Size size) {
    ///
  }
}


maze_builder/example/lib/main.dart
import 'dart:math';

import 'package:example/maze_driver.dart';
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';
import 'package:maze_builder/maze_builder.dart';
import 'dart:ui' as ui;

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> with TickerProviderStateMixin {
  List<List<Cell>> maze = [];
  late AnimationController _controller;
  BoxConstraints? viewportConstraints;

  @override
  void initState() {
    super.initState();

    _controller =
        AnimationController(vsync: this, duration: const Duration(seconds: 1));

    SchedulerBinding.instance.addPostFrameCallback((_) {
      _controller.repeat();
      generateMaze();
    });
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  /// Generate the maze
  void generateMaze() {
    int size = 24;
    Random rand = Random();
    //final stopwatch = Stopwatch()..start();
    List<List<Cell>> localMaze = generate(
        width: size, height: size, closed: true, seed: rand.nextInt(100000));

    setState(() {
      maze = localMaze;
      //mazeSolution = solution;
    });

    //print('generate maze executed in ${stopwatch.elapsed}');
  }

  void randomize() {
    generateMaze();
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
        backgroundColor: const ui.Color.fromARGB(255, 0, 0, 0),
        bottomNavigationBar: Padding(
          padding: const EdgeInsets.all(8.0),
          child: Container(
              padding: const EdgeInsets.only(left: 8.0, right: 8.0),
              decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(50.0),
                  color: const ui.Color.fromARGB(255, 0, 0, 0)),
              child: Padding(
                padding: const EdgeInsets.all(8.0),
                child: Column(
                  mainAxisSize: MainAxisSize.min,
                  children: <Widget>[
                    Row(
                      mainAxisAlignment: MainAxisAlignment.spaceBetween,
                      children: const <Widget>[],
                    ),
                  ],
                ),
              )),
        ),
        body: LayoutBuilder(builder:
            (BuildContext context, BoxConstraints viewportConstraints) {
          this.viewportConstraints = viewportConstraints;
          return Stack(children: [
            Transform.translate(
              offset: const Offset(50, 100),
              child: RepaintBoundary(
                  child: CustomPaint(
                size: const ui.Size(200, 400),
                key: UniqueKey(),
                isComplex: true,
                painter: MazeDriverCanvas(
                  controller: _controller,
                  maze: maze,
                  blockSize: 16,

                  //solution: this.mazeSolution,
                  width: viewportConstraints.maxWidth,
                  height: viewportConstraints.maxHeight,
                ),
                child: Container(
                    constraints: BoxConstraints(
                        maxWidth: viewportConstraints.maxWidth,
                        maxHeight: viewportConstraints.maxHeight)),
              )),
            ),
            Positioned(
              top: viewportConstraints.maxHeight - 100,
              left: viewportConstraints.maxWidth - 100,
              child: FloatingActionButton(
                onPressed: randomize,
                backgroundColor: Colors.green,
                child: const Icon(Icons.refresh),
              ),
            ),
          ]);

          //);
        }));
  }
}
